File: shared/src/commonMain/kotlin/com/banglu/engine/dictionary/TypingHabitNormalizer.kt
package com.banglu.engine.dictionary

/**
 * Expands non-standard phonetic input to canonical forms that may exist in the dictionary.
 *
 * Bengali typists use varied transliterations. This normalizer generates
 * alternative spellings so the dictionary can find matches.
 *
 * Example: "biggan" -> ["biggan", "bigyan", "bignon"]
 *          "dunia"  -> ["dunia", "duniya"]
 *          "shokal" -> ["shokal", "sokal"]
 */
object TypingHabitNormalizer {

    private data class HabitRule(
        val pattern: String,
        val replacements: List<String>,
        val context: String,
        val priority: Int
    )

    private val rules: List<HabitRule> = listOf(
        HabitRule("gg", listOf("gy", "gn"), "medial", 90),
        HabitRule("dd", listOf("dy", "ddh", "dv"), "medial", 90),
        HabitRule("ia", listOf("iya"), "medial", 85),
        HabitRule("nn", listOf("ny", "nno"), "medial", 80),
        HabitRule("ua", listOf("uya", "owa"), "medial", 80),
        HabitRule("ie", listOf("iye"), "medial", 80),
        HabitRule("ue", listOf("uye"), "medial", 75),
        HabitRule("io", listOf("iyo"), "medial", 75),
        HabitRule("cc", listOf("cch", "chy"), "medial", 70),
        HabitRule("bb", listOf("by", "bv"), "medial", 70),
        HabitRule("sh", listOf("s"), "initial", 60),
        HabitRule("pp", listOf("py", "pn"), "medial", 60),
        HabitRule("mm", listOf("my", "mn"), "medial", 60),
        HabitRule("ng", listOf("nk", "nkh"), "medial", 40)
    ).sortedByDescending { it.priority }

    /**
     * Expand user input into canonical phonetic variants.
     *
     * @param input The user's phonetic input
     * @param maxVariants Maximum number of variants to return (including original)
     * @return List of phonetic variants, always including the original input
     */
    fun expand(input: String, maxVariants: Int = 8): List<String> {
        val key = input.lowercase().trim()

        // Short inputs: no expansion
        if (key.isEmpty() || key.length < 3) {
            return listOf(key.ifEmpty { input })
        }

        val results = mutableSetOf(key)

        for (rule in rules) {
            if (results.size >= maxVariants) break

            // Find all occurrences of the pattern
            var idx = key.indexOf(rule.pattern)
            while (idx >= 0 && results.size < maxVariants) {
                // Check context
                val contextMatch = when (rule.context) {
                    "initial" -> idx == 0
                    "medial" -> idx > 0
                    else -> true
                }

                if (contextMatch) {
                    for (replacement in rule.replacements) {
                        if (results.size >= maxVariants) break

                        val variant = key.substring(0, idx) + replacement + key.substring(idx + rule.pattern.length)

                        if (variant.isNotEmpty() && variant != key && variant.length >= 2) {
                            results.add(variant)
                        }
                    }
                }

                // Find next occurrence after this one
                idx = key.indexOf(rule.pattern, idx + 1)
            }
        }

        return results.toList()
    }
}

File: ios-keyboard-engine/Sources/BangluKeyboardEngine/TypingHabitNormalizer.swift
import Foundation

/// Expands non-standard user input to canonical phonetic forms.
/// e.g. "gg" -> "gy", "ia" -> "iya", "sh" -> "s" (initial only)
public enum TypingHabitNormalizer {

    private struct Rule {
        let pattern: String
        let replacements: [String]
        let context: String  // "initial" or "medial"
        let priority: Int
    }

    private static let rules: [Rule] = [
        Rule(pattern: "gg", replacements: ["gy", "gn"], context: "medial", priority: 90),
        Rule(pattern: "dd", replacements: ["dy", "ddh", "dv"], context: "medial", priority: 90),
        Rule(pattern: "ia", replacements: ["iya"], context: "medial", priority: 85),
        Rule(pattern: "nn", replacements: ["ny", "nno"], context: "medial", priority: 80),
        Rule(pattern: "ua", replacements: ["uya", "owa"], context: "medial", priority: 80),
        Rule(pattern: "ie", replacements: ["iye"], context: "medial", priority: 80),
        Rule(pattern: "ue", replacements: ["uye"], context: "medial", priority: 75),
        Rule(pattern: "io", replacements: ["iyo"], context: "medial", priority: 75),
        Rule(pattern: "cc", replacements: ["cch", "chy"], context: "medial", priority: 70),
        Rule(pattern: "bb", replacements: ["by", "bv"], context: "medial", priority: 70),
        Rule(pattern: "sh", replacements: ["s"], context: "initial", priority: 60),
        Rule(pattern: "pp", replacements: ["py", "pn"], context: "medial", priority: 60),
        Rule(pattern: "mm", replacements: ["my", "mn"], context: "medial", priority: 60),
        Rule(pattern: "ng", replacements: ["nk", "nkh"], context: "medial", priority: 40),
    ].sorted { $0.priority > $1.priority }

    public static func expand(_ input: String, maxVariants: Int = 8) -> [String] {
        let key = input.lowercased().trimmingCharacters(in: .whitespaces)
        guard key.count >= 3 else { return [key] }

        var results: Set<String> = [key]
        for rule in rules {
            guard results.count < maxVariants else { break }
            guard let range = key.range(of: rule.pattern) else { continue }
            let idx = key.distance(from: key.startIndex, to: range.lowerBound)
            if rule.context == "initial" && idx != 0 { continue }
            if rule.context == "medial" && idx == 0 { continue }

            for replacement in rule.replacements {
                guard results.count < maxVariants else { break }
                let variant = key.replacingCharacters(in: range, with: replacement)
                if !variant.isEmpty && variant != key && variant.count >= 2 {
                    results.insert(variant)
                }
            }
        }
        return Array(results)
    }
}
